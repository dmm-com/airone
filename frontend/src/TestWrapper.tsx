import { createTheme, ThemeProvider } from "@mui/material/styles";
import { SnackbarProvider } from "notistack";
import React, { FC } from "react";
import { MemoryRouter } from "react-router-dom";

export const TestWrapper: FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const theme = createTheme();
  return (
    <ThemeProvider theme={theme}>
      <SnackbarProvider>
        <MemoryRouter>{children}</MemoryRouter>
      </SnackbarProvider>
    </ThemeProvider>
  );
};
