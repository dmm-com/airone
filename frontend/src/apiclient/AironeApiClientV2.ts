import Cookies from "js-cookie";

import {
  ACL,
  AclApi,
  Configuration,
  Entity,
  EntityApi,
  GetEntry,
  EntryApi,
  Group,
  GroupApi,
  Job,
  JobApi,
} from "apiclient/autogenerated";

// Get CSRF Token from Cookie set by Django
// see https://docs.djangoproject.com/en/3.2/ref/csrf/
function getCsrfToken(): string {
  return Cookies.get("csrftoken");
}

/**
 * A rich API client with using auto-generated client with openapi-generator.
 */
class AironeApiClientV2 {
  private acl: AclApi;
  private entity: EntityApi;
  private entry: EntryApi;
  private group: GroupApi;
  private job: JobApi;

  constructor() {
    const config = new Configuration({ basePath: "" });
    this.acl = new AclApi(config);
    this.entity = new EntityApi(config);
    this.entry = new EntryApi(config);
    this.group = new GroupApi(config);
    this.job = new JobApi(config);
  }

  async getAcl(id: number): Promise<ACL> {
    return this.acl.aclApiV2AclsRetrieve({ id });
  }

  async updateAcl(
    id: number,
    name: string,
    objectType: number,
    isPublic: boolean,
    defaultPermission: number,
    acl: any[]
  ): Promise<void> {
    await this.acl.aclApiV2AclsUpdate(
      {
        id,
        aCL: {
          id: id,
          name: name,
          isPublic: isPublic,
          defaultPermission: defaultPermission,
          objtype: objectType,
          acl: acl,
          // readonly
          parent: undefined,
          acltypes: undefined,
          members: undefined,
        },
      },
      {
        headers: {
          "Content-Type": "application/json;charset=utf-8",
          "X-CSRFToken": getCsrfToken(),
        },
      }
    );
  }

  async getEntities(): Promise<Entity[]> {
    return await this.entity.entityApiV2EntitiesList();
  }

  async getEntity(id: number): Promise<Entity> {
    return await this.entity.entityApiV2EntitiesRetrieve({ id });
  }

  async getEntry(id: number): Promise<GetEntry> {
    return await this.entry.entryApiV2Retrieve({ id });
  }

  async getGroups(): Promise<Group[]> {
    return await this.group.groupApiV2GroupsList();
  }

  async getGroup(id: number): Promise<Group> {
    return await this.group.groupApiV2GroupsRetrieve({ id });
  }

  async getJobs(noLimit = false): Promise<Job[]> {
    const page = await this.job.jobApiV2JobsList(noLimit ? {} : { limit: 50 });
    return page.results ?? [];
  }
}

export const aironeApiClientV2 = new AironeApiClientV2();
