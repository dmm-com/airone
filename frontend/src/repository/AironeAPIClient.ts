import Cookies from "js-cookie";

// Get CSRF Token from Cookie set by Django
// see https://docs.djangoproject.com/en/3.2/ref/csrf/
function getCsrfToken(): string {
  return Cookies.get("csrftoken") ?? "";
}

export function postLogin(formData: FormData): Promise<Response> {
  return fetch(`/auth/login/?next=${formData.get("next")}`, {
    method: "POST",
    headers: {
      "X-CSRFToken": getCsrfToken(),
    },
    body: formData,
    redirect: "manual",
  });
}

export function postLogout(): Promise<Response> {
  return fetch("/auth/logout/", {
    method: "POST",
    headers: {
      "X-CSRFToken": getCsrfToken(),
    },
  });
}
