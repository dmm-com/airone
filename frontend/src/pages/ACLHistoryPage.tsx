import { Box, Container } from "@mui/material";
import React, { FC, useEffect, useState } from "react";
import { useAsync } from "react-use";

import { ACLHistoryList } from "../components/acl/ACLHistoryList";
import { PageHeader } from "../components/common/PageHeader";
import { EntityBreadcrumbs } from "../components/entity/EntityBreadcrumbs";
import { EntryBreadcrumbs } from "../components/entry/EntryBreadcrumbs";
import { useTypedParams } from "../hooks/useTypedParams";
import { DjangoContext } from "../services/DjangoContext";

import { EntityDetail, EntryRetrieve } from "apiclient/autogenerated";
import { Loading } from "components/common/Loading";
import { aironeApiClientV2 } from "repository/AironeApiClientV2";

export const ACLHistoryPage: FC = () => {
  const djangoContext = DjangoContext.getInstance();
  const { objectId } = useTypedParams<{ objectId: number }>();
  const [entity, setEntity] = useState<EntityDetail>();
  const [entry, setEntry] = useState<EntryRetrieve>();

  const acl = useAsync(async () => {
    return await aironeApiClientV2.getAcl(objectId);
  }, [objectId]);

  const aclHistory = useAsync(async () => {
    return await aironeApiClientV2.getAclHistory(objectId);
  }, [objectId]);

  useEffect(() => {
    switch (acl.value?.objtype) {
      case djangoContext?.aclObjectType.entity:
        aironeApiClientV2.getEntity(objectId).then((resp) => {
          setEntity(resp);
        });
        break;
      case djangoContext?.aclObjectType.entry:
        aironeApiClientV2.getEntry(objectId).then((resp) => {
          setEntry(resp);
        });
        break;
    }
  }, [acl]);

  return (
    <Box className="container-fluid">
      {entry != null && <EntryBreadcrumbs entry={entry} title="ACL変更履歴" />}
      {entity != null && (
        <EntityBreadcrumbs entity={entity} title="ACL変更履歴" />
      )}

      <PageHeader title={acl.value?.name ?? ""} description="ACL変更履歴" />

      {aclHistory.loading ? (
        <Loading />
      ) : (
        <Container>
          <ACLHistoryList histories={aclHistory.value ?? []} />
        </Container>
      )}
    </Box>
  );
};
