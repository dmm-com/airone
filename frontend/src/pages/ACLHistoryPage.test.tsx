/**
 * @jest-environment jsdom
 */

import { ACLHistory } from "@dmm-com/airone-apiclient-typescript-fetch";
import {
  render,
  waitForElementToBeRemoved,
  screen,
} from "@testing-library/react";
import React from "react";

import { ACLHistoryPage } from "./ACLHistoryPage";

import { TestWrapper } from "TestWrapper";

afterEach(() => {
  jest.clearAllMocks();
});

test("should match snapshot", async () => {
  Object.defineProperty(window, "django_context", {
    value: {
      user: {
        is_superuser: false,
      },
    },
    writable: false,
  });

  const history: Array<ACLHistory> = [];
  const acl = {
    name: "acl1",
    objtype: "type1",
    is_public: false,
    default_permission: 1,
    parent: {
      id: 10,
      name: "Entity1",
    },
    acltypes: [
      {
        id: 1,
        name: "type1",
      },
    ],
    roles: [
      {
        id: 1,
        name: "member1",
        type: 1,
        current_permission: 1,
      },
    ],
  };

  /* eslint-disable */
  jest
    .spyOn(
      require("../repository/AironeApiClientV2").aironeApiClientV2,
      "getAclHistory"
    )
    .mockResolvedValue(Promise.resolve(history));
  jest
    .spyOn(
      require("../repository/AironeApiClientV2").aironeApiClientV2,
      "getAcl"
    )
    .mockResolvedValue(Promise.resolve(acl));
  /* eslint-enable */

  // wait async calls and get rendered fragment
  const result = render(<ACLHistoryPage />, {
    wrapper: TestWrapper,
  });
  await waitForElementToBeRemoved(screen.getByTestId("loading"));

  expect(result).toMatchSnapshot();
});
