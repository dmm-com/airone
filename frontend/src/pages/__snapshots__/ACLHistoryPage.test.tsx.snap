// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container-fluid MuiBox-root css-0"
      >
        <div
          class="MuiBox-root css-1wo2fgg"
        >
          <div
            class="MuiBox-root css-t1gim7"
          >
            <div
              class="MuiBox-root css-1pj6r2l"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-178toxd-MuiTypography-root"
                id="title"
              >
                acl1
              </h6>
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
                id="description"
              >
                ACL変更履歴
              </h6>
              <div
                class="MuiBox-root css-1b3gh38"
              />
            </div>
            <hr
              class="MuiDivider-root MuiDivider-fullWidth MuiDivider-flexItem css-yk10cl-MuiDivider-root"
            />
          </div>
        </div>
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <table
              class="MuiTable-root css-rqglhn-MuiTable-root"
            >
              <thead
                class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head css-15pgmhf-MuiTableRow-root"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-flimy0-MuiTableCell-root"
                    scope="col"
                  >
                    変更内容
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-flimy0-MuiTableCell-root"
                    scope="col"
                  >
                    実行日時
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-flimy0-MuiTableCell-root"
                    scope="col"
                  >
                    実行者
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
              />
            </table>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container-fluid MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-1wo2fgg"
      >
        <div
          class="MuiBox-root css-t1gim7"
        >
          <div
            class="MuiBox-root css-1pj6r2l"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-178toxd-MuiTypography-root"
              id="title"
            >
              acl1
            </h6>
            <h6
              class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
              id="description"
            >
              ACL変更履歴
            </h6>
            <div
              class="MuiBox-root css-1b3gh38"
            />
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth MuiDivider-flexItem css-yk10cl-MuiDivider-root"
          />
        </div>
      </div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <table
            class="MuiTable-root css-rqglhn-MuiTable-root"
          >
            <thead
              class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-15pgmhf-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-flimy0-MuiTableCell-root"
                  scope="col"
                >
                  変更内容
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-flimy0-MuiTableCell-root"
                  scope="col"
                >
                  実行日時
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-flimy0-MuiTableCell-root"
                  scope="col"
                >
                  実行者
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            />
          </table>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
